src/transformers/models/canine/tokenization_canine.py

```diff
@@ -0,0 +1,239 @@
+            model_max_length=model_max_length,
+            **kwargs,
+        )
+
+        # Creates a mapping for looking up the IDs of special symbols.
+        self._special_codepoints: Dict[Text, int] = {}
+        for codepoint, name in SPECIAL_CODEPOINTS.items():
+            self._special_codepoints[name] = codepoint
+
+        # Creates a mapping for looking up the string forms of special symbol IDs.
+        self._special_codepoint_strings: Dict[int, Text] = {
+            codepoint: name for name, codepoint in self._special_codepoints.items()
+        }
+
+        self._unicode_vocab_size = 1114112  # Unicode defines 1,114,112 total “codepoints”
+        self._num_special_tokens = len(self._special_codepoints)
+
+    @property
+    def vocab_size(self):
+        return self._unicode_vocab_size + self._num_special_tokens
+
+    def _tokenize(self, text: str) -> List[str]:
+        """Tokenize a string (i.e. perform character splitting)."""
+
+        pieces = [x for x in text]
+
+        return pieces
+
+    def _convert_token_to_id(self, token: str):
+        """Converts a token (i.e. a Unicode character) in an id (i.e. its integer Unicode code point value)."""
+
+        if token in self._special_codepoints:
+            return self._special_codepoints[token]
+        try:
+            return ord(token)
+        except TypeError:
+            raise ValueError(f"invalid token: '{token}'")
+
+    def _convert_id_to_token(self, index: int):
+        """Converts an integer Unicode code point (integer) in a token (str)."""
+
+        if index in self._special_codepoint_strings:
+            return self._special_codepoint_strings[index]
+        try:
+            return chr(index)
+        except TypeError:
+            raise ValueError(f"invalid id: {index}")
+
+    def convert_tokens_to_string(self, tokens):
+        out_string = "".join(tokens).strip()
+        return out_string
+
+    def build_inputs_with_special_tokens(
+        self, token_ids_0: List[int], token_ids_1: Optional[List[int]] = None
+    ) -> List[int]:
+        """
+        Build model inputs from a sequence or a pair of sequence for sequence classification tasks by concatenating and
+        adding special tokens. A CANINE sequence has the following format:
+
+        - single sequence: ``[CLS] X [SEP]``
+        - pair of sequences: ``[CLS] A [SEP] B [SEP]``
+
+        Args:
+            token_ids_0 (:obj:`List[int]`):
+                List of IDs to which the special tokens will be added.
+            token_ids_1 (:obj:`List[int]`, `optional`):
+                Optional second list of IDs for sequence pairs.
+
+        Returns:
+            :obj:`List[int]`: List of `input IDs <../glossary.html#input-ids>`__ with the appropriate special tokens.
+        """
+        sep = [self.sep_token_id]
+        cls = [self.cls_token_id]
+        if token_ids_1 is None:
+            return cls + token_ids_0 + sep
+        return cls + token_ids_0 + sep + token_ids_1 + sep
```

Comment:
Rather than duplicating the `cls + token_ids_0 + sep`, maybe do this?:

```py
result = cls + token_ids_0 + sep
if token_ids_1 is not None:
  result += token_ids_1 + sep
return result
```